src/test/ui/or-patterns/search-via-bindings.rs

```diff
@@ -0,0 +1,33 @@
+// Check that we expand multiple or-patterns from left to right.
+
+// run-pass
+
+#![feature(or_patterns)]
+#![allow(unreachable_patterns)] // FIXME(or-patterns) this shouldn't trigger
+
+fn search(target: (bool, bool, bool)) -> u32 {
+    let x = ((false, true), (false, true), (false, true));
+    let mut guard_count = 0;
+    match x {
+        ((a, _) | (_, a), (b @ _, _) | (_, b @ _), (c @ false, _) | (_, c @ true))
+            if {
+                guard_count += 1;
+                (a, b, c) == target
+            } =>
+        {
+            guard_count
+        }
+        _ => unreachable!(),
+    }
+}
+
+fn main() {
+    assert_eq!(search((false, false, false)), 1);
+    assert_eq!(search((false, false, true)), 2);
+    assert_eq!(search((false, true, false)), 3);
+    assert_eq!(search((false, true, true)), 4);
+    assert_eq!(search((true, false, false)), 5);
+    assert_eq!(search((true, false, true)), 6);
+    assert_eq!(search((true, true, false)), 7);
+    assert_eq!(search((true, true, true)), 8);
```

Comment:
This is weird and mind-bending, but correct once you bang your head against the wall enough.

impure guard code is yucky.

Reply:
Its possible the test would be a little less mind-bending if it also included this variant on `search` ([play](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d3cb5127d369bc752dc76058f70bd831)):

```rust
fn search(target: (bool, bool, bool)) -> u32 {
    let x = ((false, true), (false, true), (false, true));
    let mut guard_count = 0;
    match x {
        // ((a, _) | (_, a), (b @ _, _) | (_, b @ _), (c @ false, _) | (_, c @ true)) |
        ((a, _), (b @ _, _), (c @ false, _)) |
        ((a, _), (b @ _, _), (_, c @ true)) |
        ((a, _), (_, b @ _), (c @ false, _)) |
        ((a, _), (_, b @ _), (_, c @ true)) |
        ((_, a), (b @ _, _), (c @ false, _)) |
        ((_, a), (b @ _, _), (_, c @ true)) |
        ((_, a), (_, b @ _), (c @ false, _)) |
        ((_, a), (_, b @ _), (_, c @ true))
            if {
                guard_count += 1;
                (a, b, c) == target
            } =>
        {
            guard_count
        }
        _ => unreachable!(),
    }
}
```

this is of course just the original OR-patterns now expanded into the old syntax. Seeing that made me immediately understand what was going on in the test.

Now, whether this language feature (mixing multi-pattern arms and guards) is an anti-pattern in disguise is another matter entirely ... 

(I'll admit that when I initially read examples like these, I wrongly assume that we commit to the pattern, check the guard, and when the guard fails, *move on* to the next **arm**. That is of course not what happens, which is probably for the best, but it definitely can make impure guards very confusing.)